File: src/scraper/crawler.ts
import { chromium } from "playwright";
import type { CrawlOptions, CrawledPage } from "./types";

const ASSET_EXTENSIONS = /\.(png|jpg|jpeg|gif|svg|webp|ico|css|js|ts|woff|woff2|ttf|eot|pdf|zip|tar|gz|mp4|mp3|wav)$/i;

export function shouldFollowUrl(
  url: string,
  seedOrigin: string,
  urlPattern?: string
): boolean {
  if (!url || url.startsWith("#") || url.startsWith("mailto:") || url.startsWith("tel:") || url.startsWith("javascript:")) {
    return false;
  }
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    return false;
  }
  if (parsed.origin !== new URL(seedOrigin).origin) return false;
  if (ASSET_EXTENSIONS.test(parsed.pathname)) return false;
  if (urlPattern && !new RegExp(urlPattern).test(parsed.pathname)) return false;
  return true;
}

export function normalizeLinks(hrefs: string[], baseUrl: string): string[] {
  const seen = new Set<string>();
  const results: string[] = [];
  for (const href of hrefs) {
    try {
      const abs = new URL(href, baseUrl);
      abs.hash = "";
      // Strip trailing slash (but keep root)
      const normalized = abs.href.replace(/\/$/, "") || abs.href;
      if (!seen.has(normalized)) {
        seen.add(normalized);
        results.push(normalized);
      }
    } catch {
      // skip unparseable
    }
  }
  return results;
}

export function filterLinks(
  urls: string[],
  seedOrigin: string,
  urlPattern?: string
): string[] {
  return urls.filter((u) => shouldFollowUrl(u, seedOrigin, urlPattern));
}

export async function* crawl(options: CrawlOptions): AsyncGenerator<CrawledPage> {
  const {
    seedUrl,
    maxDepth,
    maxPages,
    concurrency,
    delayMs,
    urlPattern,
    waitUntil,
    selector: _selector,
  } = options;

  const browser = await chromium.launch({ headless: true });
  const visited = new Set<string>();
  const queue: { url: string; depth: number }[] = [{ url: seedUrl, depth: 0 }];
  let pageCount = 0;

  const processUrl = async (url: string, depth: number): Promise<CrawledPage | null> => {
    const page = await browser.newPage();
    try {
      await page.goto(url, { waitUntil, timeout: 15000 });
      const html = await page.content();
      const title = await page.title();
      const hrefs = await page.$$eval("a[href]", (els) =>
        els.map((el) => (el as HTMLAnchorElement).href)
      );
      await page.close();

      const normalized = normalizeLinks(hrefs, url);
      const links = filterLinks(normalized, seedUrl, urlPattern);

      return { url, html, title, links, depth };
    } catch (err) {
      await page.close().catch(() => {});
      console.error(`[crawler] Error fetching ${url}:`, err instanceof Error ? err.message : err);
      return null;
    }
  };

  try {
    while (queue.length > 0 && pageCount < maxPages) {
      const currentLevel = queue.splice(0, queue.length);
      const toProcess = currentLevel.filter(({ url }) => {
        if (visited.has(url)) return false;
        visited.add(url);
        return true;
      });

      // Process this level with concurrency chunks
      const chunks: typeof toProcess[] = [];
      for (let i = 0; i < toProcess.length; i += concurrency) {
        chunks.push(toProcess.slice(i, i + concurrency));
      }

      for (const chunk of chunks) {
        if (pageCount >= maxPages) break;
        const settled = await Promise.allSettled(
          chunk.map(({ url, depth }) => processUrl(url, depth))
        );
        for (const result of settled) {
          if (pageCount >= maxPages) break;
          if (result.status === "fulfilled" && result.value) {
            const crawledPage = result.value;
            pageCount++;
            yield crawledPage;

            // Queue next depth
            if (crawledPage.depth < maxDepth) {
              for (const link of crawledPage.links) {
                if (!visited.has(link) && pageCount < maxPages) {
                  queue.push({ url: link, depth: crawledPage.depth + 1 });
                }
              }
            }
          }
        }
        if (delayMs > 0 && queue.length > 0) {
          await new Promise((r) => setTimeout(r, delayMs));
        }
      }
    }
  } finally {
    await browser.close();
  }
}

File: mairu/internal/scraper/crawler.go
package scraper

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/PuerkitoBio/goquery"
)

var assetExtensions = regexp.MustCompile(`(?i)\.(png|jpg|jpeg|gif|svg|webp|ico|css|js|ts|woff|woff2|ttf|eot|pdf|zip|tar|gz|mp4|mp3|wav)$`)

func shouldFollowURL(testURL, seedOrigin, urlPattern string) bool {
	if testURL == "" || strings.HasPrefix(testURL, "#") || strings.HasPrefix(testURL, "mailto:") || strings.HasPrefix(testURL, "tel:") || strings.HasPrefix(testURL, "javascript:") {
		return false
	}
	parsed, err := url.Parse(testURL)
	if err != nil {
		return false
	}
	seedParsed, err := url.Parse(seedOrigin)
	if err != nil {
		return false
	}
	if parsed.Host != "" && parsed.Host != seedParsed.Host {
		return false
	}
	if assetExtensions.MatchString(parsed.Path) {
		return false
	}
	if urlPattern != "" {
		matched, _ := regexp.MatchString(urlPattern, parsed.Path)
		if !matched {
			return false
		}
	}
	return true
}

func normalizeLinks(hrefs []string, baseURL string) []string {
	seen := make(map[string]bool)
	var results []string
	base, err := url.Parse(baseURL)
	if err != nil {
		return nil
	}

	for _, href := range hrefs {
		parsed, err := url.Parse(href)
		if err != nil {
			continue
		}
		abs := base.ResolveReference(parsed)
		abs.Fragment = ""
		normalized := abs.String()
		if strings.HasSuffix(normalized, "/") && len(normalized) > 1 {
			normalized = strings.TrimSuffix(normalized, "/")
		}
		if !seen[normalized] {
			seen[normalized] = true
			results = append(results, normalized)
		}
	}
	return results
}

func filterLinks(urls []string, seedOrigin, urlPattern string) []string {
	var results []string
	for _, u := range urls {
		if shouldFollowURL(u, seedOrigin, urlPattern) {
			results = append(results, u)
		}
	}
	return results
}

func fetchPage(targetURL string) (*CrawledPage, error) {
	client := &http.Client{Timeout: 15 * time.Second}
	req, err := http.NewRequest("GET", targetURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", "mairu-crawler/1.0")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("status code %d", resp.StatusCode)
	}

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	html := string(bodyBytes)

	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		return nil, err
	}

	title := doc.Find("title").Text()
	var hrefs []string
	doc.Find("a[href]").Each(func(i int, s *goquery.Selection) {
		if href, exists := s.Attr("href"); exists {
			hrefs = append(hrefs, href)
		}
	})

	return &CrawledPage{
		URL:   targetURL,
		HTML:  html,
		Title: title,
		Links: filterLinks(normalizeLinks(hrefs, targetURL), targetURL, ""),
	}, nil
}

// Crawl performs a breadth-first search of the given seed URL.
// Since Go doesn't have async generators like TS, we use channels.
func Crawl(options CrawlOptions, out chan<- CrawledPage) {
	defer close(out)

	visited := make(map[string]bool)
	type queueItem struct {
		url   string
		depth int
	}
	queue := []queueItem{{url: options.SeedURL, depth: 0}}
	pageCount := 0

	for len(queue) > 0 && pageCount < options.MaxPages {
		currentLevel := queue
		queue = nil

		var toProcess []queueItem
		for _, item := range currentLevel {
			if !visited[item.url] {
				visited[item.url] = true
				toProcess = append(toProcess, item)
			}
		}

		concurrency := options.Concurrency
		if concurrency <= 0 {
			concurrency = 1
		}

		for i := 0; i < len(toProcess); i += concurrency {
			if pageCount >= options.MaxPages {
				break
			}
			end := i + concurrency
			if end > len(toProcess) {
				end = len(toProcess)
			}
			chunk := toProcess[i:end]

			var wg sync.WaitGroup
			results := make([]*CrawledPage, len(chunk))
			for j, item := range chunk {
				wg.Add(1)
				go func(idx int, qItem queueItem) {
					defer wg.Done()
					page, err := fetchPage(qItem.url)
					if err == nil && page != nil {
						page.Depth = qItem.depth
						results[idx] = page
					}
				}(j, item)
			}
			wg.Wait()

			for _, page := range results {
				if page == nil || pageCount >= options.MaxPages {
					continue
				}
				page.Links = filterLinks(page.Links, options.SeedURL, options.URLPattern)
				pageCount++
				out <- *page

				if page.Depth < options.MaxDepth {
					for _, link := range page.Links {
						if !visited[link] && pageCount < options.MaxPages {
							queue = append(queue, queueItem{url: link, depth: page.Depth + 1})
						}
					}
				}
			}

			if options.DelayMs > 0 && len(queue) > 0 {
				time.Sleep(time.Duration(options.DelayMs) * time.Millisecond)
			}
		}
	}
}
